File: acmicpc/A01654.py
from sys import stdin as s


def binary_search(cables, len_cables, n):
    low, high = 0, 2_147_483_648 - 1
    while low <= high:
        length = (low + high) // 2
        count = 0
        for cable in cables:
            count += cable // length
        if count < n:
            high = length - 1
        else:
            low = length + 1
    print(high)



def reader():
    k, n = map(int, s.readline().split())
    cables = [int(s.readline().strip()) for _ in range(k)]
    binary_search(cables, k, n)


if __name__ == '__main__':
    reader()

File: acmicpc/A01654.java
package acmicpc;

import java.io.*;
import java.util.*;

public class A01654 {
    public static long solution(int[] cables, int num_cables, int required) {
        long low = 0L, high = 2_147_483_648L - 1;
        while (low <= high) {
            long length = low + (high - low) / 2, count = 0L;
            for (int cable : cables)
                count += cable / length;
            if (count < required)
                high = length - 1L;
            else
                low = length + 1L;
        }
        return high;
    }

    public void reader() throws NumberFormatException, IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int k = Integer.parseInt(st.nextToken()), n = Integer.parseInt(st.nextToken());
        int[] cables = new int[k];
        for (int i = 0; i < k; i++)
            cables[i] = Integer.parseInt(br.readLine());
        System.out.println(solution(cables, k, n));
    }

    public static void main(String[] args) throws NumberFormatException, IOException {
        A01654 test = new A01654();
        test.reader();
    }
}
